### DISCUSSION TITLE
[RFC] Table 支持可伸缩列

### DISCUSSION THREAD

**Discussion by linxianxi** [ant-design/ant-design#48152] [RFCs]
## Summary
Table 底层支持可伸缩列，否则结合第三方库（例如 react-resizable） 会有一些问题

## Motivation
rc-table 使用 colgroup 标签控制列宽，原生 colgroup 标签会对 width 做一些特殊处理，虽然挂在标签上的 width 是用户设置的 width，但和实际渲染出来的宽度是不同的。

### 现在的情况
```
// width 800px
<Table columns={[{width: 100}, {width: 100}]}  />
```
但实际宽度是 [400px, 400px]
![image](https://github.com/ant-design/ant-design/assets/47104575/07fbb4ac-e7da-4fc9-99b8-435208558621)

## Detailed design

开启 resizable 时， tableLayout 为 `fixed`（因为列增大时，其它任何列不应该被减小宽度）。colgroup 身上 style 的 width 为真实宽度，并且 column width 限制为 number 类型，不设置的话给个默认宽度。

实现方式：column width 受控

```
// width 800px
<Table columns={[{width: 100}, {width: 100}]}  />
```
![image](https://github.com/ant-design/ant-design/assets/47104575/a86419ee-5932-4c3c-b79a-2bf03c2332e2)

### Table

| prop | type | description |
| --- | --- | --- |
|  columnResizable  |  boolean \| { showLine?: boolean } | 开启 |
| onColumnResize |  (info: {columnKey: React.Key, width: number, columnKeyWidths?: { columnKey: React.Key; width: number } [] }) => void;  | 暂不提供 |
| onColumnResizeStart | (columnKey: React.Key) => void | （暂不提供） |
| onColumnResizeComplete | 同 onColumnResize | 结束 |

onColumnResize 返回的 width 是真实的宽度 px, 用户如果想保存到 localStorage 应该保存全部列的真实宽度。因为应使用真实的宽度 px 进行保存，否则像 `columns=[{width: 100}, {width: 100}]`，onResize 后，得到一列是真实的，一列还是假的。

### Column

| prop | type |
| --- | ---- |
|  resizable  |  boolean \| { minWidth?: number } |

#### 1、设定的列宽小于 Table 的宽度时
```
// width 800px
<Table columns={[{width: 100}, {width: 100}]} />
```
实际宽度为 [400px, 400px]，当缩小第一列到 300 时，两列宽度加起来小于了表格的宽度 800 ，第二列应该变为 500，保持不小于 800。这样 onColumnResizeComplete 中的 columnKeyWidths 分别为 300px 和 500px。

......

**Comment by crazyair** [comment]
这个可是收费功能(某些库)

**Comment by Veveue** [comment]
```js

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    key: 'name',
    onCell: () => { // 单元格属性
      style: { }
    },
    hidden: true // 是否渲染列
  }]
```
试试

**Reply by linxianxi** [reply]
什么意思？

**Comment by MadCcc** [comment]
onResize -> onColumnResize

**Reply by MadCcc** [reply]
三个参数合成一个对象

**Reply by MadCcc** [reply]
未超出容器宽度时，总宽度不变

**Reply by MadCcc** [reply]
mode 不需要。用户受控处理

**Reply by MadCcc** [reply]
onResizeEnd -> onColumnResizeComplete
onResizeStart 先不做

**Reply by linxianxi** [reply]
未超出容器宽度时，缩小时，总宽度不变。放大时，总宽度应该跟着变大

**Comment by crazyair** [comment]
onResizeChange

**Comment by crazyair** [comment]
```tsx
resizable:{
    minWidth:
    maxWidth: 
    onChange: (type: start | end,dataIndex )
}
```

**Comment by liixing** [comment]
目前配合 react-resize 会导致 头部筛选失效， columns 的 title 额外渲染的元素也会有问题
